src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts

```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
         this.grid.selectRow(row, debounce);
     }
 
+    hasNodeWithDataId(dataId: string): boolean {
+        return !!this.root.getNodeByDataId(dataId);
+    }
+
+    protected insertDataToParentNode(data: DATA, parent: TreeNode<DATA>, index: number) {
+        const nodeToInsert: TreeNode<DATA> = this.dataToTreeNode(data, parent);
+        this.insertNodeToParentNode(nodeToInsert, parent, index);
+    }
+
+    protected appendDataToParentNode(data: DATA, parent: TreeNode<DATA>) {
+        const insertIndex: number = parent.getChildren().length;
+        this.insertDataToParentNode(data, parent, insertIndex);
+    }
+
+    protected insertNodeToParentNode(nodeToInsert: TreeNode<DATA>, parent: TreeNode<DATA>, index: number) {
+        parent.insertChild(nodeToInsert, index);
+        parent.setExpandable(true);
+
+        const parentRow: number = this.gridData.getRowById(parent.getId());
+        if (!!parentRow || parentRow === 0) {
+            this.gridData.insertItem(parentRow + index + 1, nodeToInsert);
+        }
+
+        this.invalidateNodes([parent]);
+    }
+
+    moveNode(from: number, to: number): number {
+        const root: TreeNode<DATA> = this.getRoot().getCurrentRoot();
+        const rootChildren: TreeNode<DATA>[] = root.treeToList();
+
+        const item: TreeNode<DATA> = rootChildren.slice(from, from + 1)[0];
+        rootChildren.splice(rootChildren.indexOf(item), 1);
+        rootChildren.splice(to, 0, item);
+
+        this.initData(rootChildren);
+        root.setChildren(rootChildren);
+
+        return rootChildren.indexOf(item);
+    }
+
+    getDataByRow(row: number): DATA {
+        const nodes: TreeNode<DATA>[] = this.getRoot().getCurrentRoot().treeToList();
+        const node = nodes[row];
+
+        return !!node ? node.getData() : null;
+    }
+
+    getParentDataById(dataId: string): DATA {
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromDefault(dataId);
+
+        if (!node) {
+            return null;
+        }
+
+        return node.getParent().getData();
+    }
+
+    getCurrentData(expanded: boolean = true): DATA[] {
+        return this.root.getCurrentRoot().treeToList(false, expanded).map((node: TreeNode<DATA>) => node.getData());
+    }
+
+    getDataLevel(data: DATA): number {
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromCurrent(data.getId());
+        return !!node ? node.calcLevel() : -1;
+    }
+
+    isExpandedAndHasChildren(dataId: string): boolean {
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromCurrent(dataId);
+
+        return !!node ? node.isExpandable() && node.hasChildren() : false;
+    }
+
+    protected isToBeExpanded(node: TreeNode<DATA>): boolean {
+        return this.expandedNodesDataIds.indexOf(node.getDataId()) > -1;
+    }
+
+    getDataFromDomEvent(e: DOMEvent): DATA {
```

Comment:
Codacy found an issue: ['DOMEvent' is not defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)